stats_overlay.js

```diff
@@ -0,0 +1,284 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @fileoverview A statistics visualiser.
+ */
+
+
+
+/** @constructor */
+shaka.StatsOverlay = function() {
+
+  /** @private {?shaka.player.Player} */
+  this.player_;
+
+  /** @private {?HTMLElement} */
+  this.canvas_;
+
+  /** @private {number} */
+  this.width_ = 200;
+
+  /** @private {number} */
+  this.height_ = 100;
+
+  /** @private {number} */
+  this.maxY_ = 1;
+
+  /** @private {?number} */
+  this.startTimestamp_;
+
+  /** @private {?number} */
+  this.currentXOffset_;
+
+  /** @private {?number} */
+  this.refreshIntervalId_;
+
+};
+
+
+/**
+ * Initializes the stats ovelay.
+ * @param {!shaka.player.Player} player
+ * @param {?HTMLElement} canvas
+ */
+shaka.StatsOverlay.prototype.init = function(player, canvas) {
+
+  this.player_ = player;
+  this.canvas_ = canvas;
+
+  this.width_ = canvas.width;
+  this.height_ = canvas.height - 5;
+
+  this.reset();
+};
+
+
+/**
+ * Converts a timestamp to an x coordinate
+ * @private
+ * @param {?number} timestamp
+ * @return {number}
+ */
+shaka.StatsOverlay.prototype.convertTimestampToXCoordinate_ =
+    function(timestamp) {
+  return Math.round(timestamp - this.startTimestamp_ + this.currentXOffset_) +
+         0.5;
+};
+
+
+/**
+ * Converts bits to the Y Coordinate
+ * Updated the running maxY_ to calculate the correct scale
+ * @private
+ * @param {?number} bits
+ * @return {number}
+ */
+shaka.StatsOverlay.prototype.convertBitsToYCoordinate_ = function(bits) {
+
+  // update the running scale
+  if (bits && bits > this.maxY_) {
+    this.maxY_ = bits;
+
+    // update the scaleY
+    this.scaleY_ = - (this.height_ / (this.maxY_ * 1.1));
+  }
+
+  // fudge the coordindate to ensure that the paths produce clean lines
+  return Math.round((bits * this.scaleY_) + this.height_) + 0.5;
+};
+
+
+/**
+ * Reset the overlay
+ */
+shaka.StatsOverlay.prototype.reset = function() {
+  this.maxY_ = 1;
+  this.scaleY_ = - (this.height_ / this.maxY_);
+};
+
+
+/**
+ * Draw stats to the overlay
+ * @private
+ */
+shaka.StatsOverlay.prototype.draw_ = function() {
+
+  var context = this.canvas_.getContext('2d');
+  var stats = this.player_.getStats();
+
+  // check if there is any band width history
+  // reset the overlay to as the video has changed
+  if (stats.bandwidthHistory.length == 0) {
+    this.reset();
+    return;
+  }
+
+  // play start time is determined by the first bandwidth history
+  this.startTimestamp_ = stats.bandwidthHistory[0].timestamp;
+
+  // shift the output based on the size of the canvas and the elaspsed time
+  this.currentXOffset_ = - Math.max(0,
+      (shaka.util.Clock.now() / 1000) - this.startTimestamp_ - this.width_);
+
+  // clear the canvas
+  this.canvas_.width = this.canvas_.width;
```

Comment:
Seems like this is not the best way to clear the canvas.  From what I can tell, clearRect() is both faster and more portable than resetting width.